system/controllers.go

```diff
@@ -0,0 +1,63 @@
+package system
+
+import (
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/ausocean/cloud/model"
+)
+
+// ControllerSystem represents a controller device which has associated
+// variables, sensors, acuators, and peripherals.
+type ControllerSystem struct {
+	Controller  model.Device
+	Variables   []*model.Variable
+	Sensors     []*model.SensorV2
+	Actuators   []*model.ActuatorV2
+	Peripherals []*model.Device
+}
+
+// Option represents functional options that can be passed to NewControllerSystem.
+type Option func(*ControllerSystem) error
+
+func NewControllerSystem(skey int64, MAC, name string, options ...Option) (*ControllerSystem, error) {
+	// Check MAC.
+	if model.MacEncode(MAC) == 0 {
+		return nil, model.ErrInvalidMACAddress
+	}
+
+	// Create a ControllerSystem.
+	sys := &ControllerSystem{
+		Controller: model.Device{
+			Skey:    skey,
+			Mac:     model.MacEncode(MAC),
+			Name:    name,
+			Type:    model.DevTypeController,
+			Enabled: true,
+		},
+	}
+
+	// Apply functional options.
+	for i, opt := range options {
+		err := opt(sys)
+		if err != nil {
+			return nil, fmt.Errorf("failed to apply option[%d]: %w", i, err)
+		}
+	}
+
+	// Append site/device information to actuators, sensors, and variables.
+	for _, variable := range sys.Variables {
+		log.Println("updating vars.")
+		variable.Skey = skey
+		variable.Scope = strings.ReplaceAll(sys.Controller.MAC(), ":", "")
+	}
+	for _, sensor := range sys.Sensors {
+		sensor.Mac = sys.Controller.Mac
```

Comment:
The sensor must have the MAC set before PutSensor is called. Is that the case?

Reply:
In the current implementation it is up to the user to put the controller system (and all components) into the datastore, so currently, yes this happens before then.

Reply:
This functionality will be in a future change I think. I agree that having all the components be abstracted as one system for datastore operations is certainly the way to go. At the moment though, I am just creating the system.